genie-web/src/main/java/com/netflix/genie/web/rpc/grpc/services/impl/v4/GRpcJobFileSyncServiceImpl.java

```diff
@@ -0,0 +1,609 @@
+                                                    file ->
+                                                        com.netflix.genie.proto.JobFileState
+                                                            .newBuilder()
+                                                            .setPath(file.getPath())
+                                                            .setSize(file.getSize())
+                                                            .build()
+                                                )
+                                                .collect(Collectors.toSet())
+                                        )
+                                        .build())
+                                .build()
+                        )
+                        .build()
+                );
+            }
+        }
+
+        private void handleDataUpload(final DataUpload dataUpload) {
+            final String messageId = dataUpload.getId();
+            log.debug("Received data upload message with id {} for job {}", messageId, this.jobId);
+
+            // Write some data and save the message in the ack buffer
+            if (this.waitingForBeginMessage) {
+                log.debug(
+                    "Haven't received a {} message. Ignoring data upload message {}.",
+                    BeginSync.class.getCanonicalName(),
+                    messageId
+                );
+                this.sendResetMessageIfNecessary();
+            } else {
+                try {
+                    this.jobFileService.updateFile(
+                        this.jobId,
+                        dataUpload.getPath(),
+                        dataUpload.getStartByte(),
+                        // TODO: There are other methods here to go right to a ByteBuffer we might want to explore
+                        dataUpload.getData().toByteArray()
+                    );
+
+                    synchronized (this.messagesLock) {
+                        this.successfulSyncIds.add(messageId);
+                    }
+                } catch (final Exception e) {
+                    // For some reason saving the log failed. Mark this as a failed message
+                    log.error(
+                        "Unable to save data for job {} from message {} due to {}",
+                        this.jobId,
+                        messageId,
+                        e.getMessage(),
+                        e
+                    );
+
+                    synchronized (this.messagesLock) {
+                        this.failedSyncIds.add(messageId);
+                    }
+                }
+
+                this.checkIfShouldSendAck();
+            }
+        }
+
+        private void handleDeleteFile(final DeleteFile deleteFile) {
+            final String messageId = deleteFile.getId();
+            log.debug("Received file delete message {} in job {}", messageId, this.jobId);
+
+            if (this.waitingForBeginMessage) {
```

Comment:
This block appears in 3 methods, could be pulled out for less duplication and to make these methods simpler.

Reply:
I left it in each due to wanting unique log messages that had specific details. Can change it later if we find it a burden. The logic to check whether to send the ack is already shared in one spot they all call.